src/Quasar.jl

```diff
@@ -0,0 +1,1917 @@
+    gate_def_visitor(raw_expressions)
+    return gate_def_visitor.instructions
+end
+
+function evaluate_unary_op(op::Symbol, arg)
+    op == :! && return !arg
+    op == :~ && return .!arg
+    op == :- && return -arg 
+end
+function evaluate_unary_op(op::Symbol, arg::BitVector)
+    op == :! && return !any(arg)
+    op == :~ && return .!arg
+end
+
+# semgrep rules can't handle this macro properly yet
+# nosemgrep
+function evaluate_binary_op(op::Symbol, @nospecialize(lhs), @nospecialize(rhs))
+    op == :< && return lhs < rhs
+    op == :> && return lhs > rhs
+    op == :<= && return lhs <= rhs
+    op == :>= && return lhs >= rhs
+    op == Symbol("=") && return rhs
+    op == Symbol("!=") && return lhs != rhs
+    op == Symbol("==") && return lhs == rhs
+    op == :+ && return lhs + rhs
+    op == :- && return lhs - rhs
+    op == :* && return lhs * rhs
+    op == :/ && return lhs / rhs
+    op == :% && return lhs % rhs
+    op == Symbol("<<") && return lhs << rhs
+    op == Symbol(">>") && return lhs >> rhs
+    op == Symbol("**") && return lhs ^ rhs
+    op == Symbol("&&") && return lhs && rhs
+    op == Symbol("||") && return lhs || rhs
+    op == :| && return lhs .| rhs
+    op == :& && return lhs .& rhs
+    op == :^ && return lhs .⊻ rhs
+end
+
+function name(expr::QasmExpression)::String
+    head(expr) == :identifier         && return expr.args[1]::String
+    head(expr) == :indexed_identifier && return name(expr.args[1]::QasmExpression)
+    head(expr) == :qubit_declaration  && return name(expr.args[1]::QasmExpression)
+    head(expr) == :classical_declaration && return name(expr.args[2]::QasmExpression)
+    head(expr) == :input              && return name(expr.args[2]::QasmExpression)
+    head(expr) == :function_call      && return name(expr.args[1]::QasmExpression)
+    head(expr) == :gate_call          && return name(expr.args[1]::QasmExpression)
+    head(expr) == :gate_definition    && return name(expr.args[1]::QasmExpression)
+    head(expr) == :classical_assignment && return name(expr.args[1].args[2]::QasmExpression)
+    throw(QasmVisitorError("name not defined for expressions of type $(head(expr))"))
+end
+
+evaluate(v::AbstractVisitor, i::Number) = i
+evaluate(v::AbstractVisitor, i::String) = i
+evaluate(v::AbstractVisitor, i::BitVector) = i
+evaluate(v::AbstractVisitor, i::NTuple{N,<:Number}) where {N} = i
+evaluate(v::AbstractVisitor, i::Vector{<:Number}) = i
+function evaluate(v::AbstractVisitor, expr::QasmExpression)
+    if head(expr) == :identifier
+        id_name = name(expr)
+        haskey(classical_defs(v), id_name) && return classical_defs(v)[id_name].val
+        haskey(qubit_mapping(v), id_name) && return evaluate_qubits(v, expr)
+    elseif head(expr) == :indexed_identifier
+        identifier_name =  name(expr)
+        if haskey(classical_defs(v), identifier_name)
+            var = classical_defs(v)[identifier_name]
+            ix  = evaluate(v, expr.args[2]::QasmExpression)
```

Comment:
Definitely not to be done in this PR, but is it possible to give names to the args for each expression type, instead of relying on indices? That can make reading much easier.

Reply:
It's kind of hard because the args are a `Vector{Any}` by design, unless you mean instantiating some variable to be `= expr.args[2]`?

Reply:
Exactly that, but only if it's done for each `Expr` subtype